qutip/core/data/expect.pxd

```diff
@@ -3,13 +3,13 @@
 
 from qutip.core.data cimport CSR, Dense, Data, Dia
 
-cpdef double complex expect_csr(CSR op, CSR state) except * nogil
-cpdef double complex expect_super_csr(CSR op, CSR state) except * nogil
+cpdef double complex expect_csr(CSR op, CSR state) except *
+cpdef double complex expect_super_csr(CSR op, CSR state) except *
 
-cpdef double complex expect_csr_dense(CSR op, Dense state) except * nogil
+cpdef double complex expect_csr_dense(CSR op, Dense state) except *
 cpdef double complex expect_super_csr_dense(CSR op, Dense state) except * nogil
```

Comment:
How come we still have `expect * nogil` here?

Reply:
It's not wrong to use `except * nogil`, it's just not optimal as cython need to get the gil to check if an error was raised when the function return. Returning a fix value for errors (`except -1`) is more efficient, as cython can check the error without the gil, but it's not supported for complex return.

For a function that is called outside another nogil block, it has no performance impact so we don't need to remove them.

Reply:
I would make these function signatures more similar and just remove `nogil` from `expect_super_csr_dense`. Otherwise the slight performance differences will be annoying.

Reply:
I did not randomly chose to skip `expect_super_csr_dense`, it use the nogil while `expect_csr_dense` call other function that run in `nogil`, but itself do not benefit from it.
Since they don't do the same computation, performance differences are expected.
And performance should be closer to each other this way as `expect_csr_dense` would go back and forth gil/nogil twice as it was instead of once like `expect_super_csr_dense`. Now it's once for both function...